Feat(digital guide)/add micronavigation info

thesun901 opened the pull request:
First of all what's added:

![image](https://github.com/user-attachments/assets/f903ab92-f137-4690-90e4-65e5e3ca22a9)
![image](https://github.com/user-attachments/assets/e6cf33bd-1275-4f46-8664-2e975358713c)


-add audioplayers in pubspec yaml
-two views above
-overall there is new endpoint!!!! so I will let you know on discord whats to add in .env
-also mentioned in readme.md


about micronavigation_response I am pretty sure that I left some redundant values but I have to check if we won't use it somewhere else so I leave it for now

mikolaj-jalocha commented on lib/features/digital_guide_view/tabs/micronavigation/data/models/micronavigation_response.dart:
```diff
@@ -0,0 +1,78 @@
+    @JsonKey(name: "name_override")
+    required MicronavigationNameOverride nameOverride,
+    @JsonKey(name: "web_content") required MicronavigationContent webContent,
+    required int sn,
```
`sn` quite mystery / ambiguous . I'd consider different name

simon-the-shark replied:
and here `sn` whatever this is can be set manually with `@JsonKey(name: "name_override")`

mikolaj-jalocha commented on lib/features/digital_guide_view/tabs/micronavigation/presentation/micronavigation_detail_view.dart:
```diff
@@ -0,0 +1,81 @@
+import "widgets/my_audio_player.dart";
+
+@RoutePage()
+class MicronavigationDetailView extends StatefulWidget {
+  const MicronavigationDetailView({
```
why `StatefulWidget` instead of `Consumer` widget and `riverpod`?

simon-the-shark replied:
does it even have state? why this isn't stateless widget?

mikolaj-jalocha commented on lib/features/digital_guide_view/tabs/micronavigation/presentation/micronavigation_expansion_tile_content.dart:
```diff
@@ -0,0 +1,84 @@
+          DigitalGuideConfig.paddingMedium,
+          0,
+        ),
+        child: DigitalGuideNavLink(
+          //CHANGE THIS WHEN WE WILL MAKE TRANSLATIONS!
```
Probably better to include that comment in TODO

simon-the-shark replied:
or just remove it, it's pretty obvious

mikolaj-jalocha commented on lib/features/digital_guide_view/tabs/micronavigation/presentation/widgets/my_audio_player.dart:
```diff
@@ -0,0 +1,105 @@
+import "../../../../../../config/ui_config.dart";
+import "../../../../../../theme/app_theme.dart";
+
+class MyAudioPlayer extends StatefulWidget {
+  final String audioUrl;
+
```
Again, can we replace that with Riverpod or Flutter Hooks maybe?

simon-the-shark replied:
flutter hooks probably better for this case, cause it's a bit more local state than global

thesun901 replied:
also I am sorry because I didnt use flutter hooks before (such a shame) so I will educate myself! After quick research its a great idea

mikolaj-jalocha commented on lib/features/digital_guide_view/tabs/micronavigation/presentation/widgets/my_audio_player.dart:
```diff
@@ -0,0 +1,105 @@
+    );
+  }
+
+  String _formatTime(Duration duration) {
+    final minutes = duration.inMinutes;
+    final seconds = duration.inSeconds % 60;
+    return "$minutes:${seconds.toString().padLeft(2, '0')}";
+  }
```
I'm not forcing this, but we have a file when all extensions related to time're stored. Cosnider that ;)

thesun901 replied:
I think I won't add that, because our time utils are connected to DateTime... and here I use Duration (changing this DateTime is bad idea, cause DateTime represents point in time) I don't think this function will be used anywhere else for now and adding it as a separate 3-liner is kind of overkill

mikolaj-jalocha commented on lib/features/digital_guide_view/tabs/micronavigation/data/models/micronavigation_response.dart:
Before merge it'd be great to clear all redundant fields. Localization, sn, inside?

mikolaj-jalocha reviewed (changes requested):
Good work here! Everything looks fine and esthetic. I've some questions to certain solutions that you've used I also left some suggestions! So for now, I request changes until  all questions will be resolved ;)

thesun901 commented:
@mikolaj-jalocha the suggestions are great, so I am not defending my solutions ;) - I will just correct my code according to your comments.

simon-the-shark commented on lib/features/digital_guide_view/data/models/digital_guide_response.dart:
```diff
@@ -7,6 +7,7 @@ part "digital_guide_response.g.dart";
 class DigitalGuideResponse with _$DigitalGuideResponse {
   const factory DigitalGuideResponse({
     required int id,
+    @JsonKey(name: "external_id") required int externalId,
```
now I've noticed that instead of manually renaming all fields, we can just use `@JsonSerializable(fieldRename: FieldRename.snake)` on the constructor and manually set only those that we want to be different

thesun901 replied:
didn't know about that but its very cool - surely will try to change that

simon-the-shark commented on lib/features/digital_guide_view/tabs/micronavigation/data/models/micronavigation_response.dart:
```diff
@@ -0,0 +1,78 @@
+    required int location,
+    @JsonKey(name: "name_override")
+    required MicronavigationNameOverride nameOverride,
+    @JsonKey(name: "web_content") required MicronavigationContent webContent,
```
same, here, we could use `@JsonSerializable(fieldRename: FieldRename.snake)`

simon-the-shark commented on lib/features/digital_guide_view/tabs/micronavigation/data/models/micronavigation_response.dart:
```diff
@@ -0,0 +1,78 @@
+
+@freezed
+class MicronavigationContent with _$MicronavigationContent {
+  const factory MicronavigationContent({
```
what;s the difference between `MicronavigationNameOverride` and `MicronavigationContent` and why this isnt a single model?

thesun901 replied:
One has Name of facility, the other has comment, I dont know really, maybe I should do it different but I they did this in such a weird manner

simon-the-shark commented on lib/features/digital_guide_view/tabs/micronavigation/data/repository/micronavigation_repository.dart:
```diff
@@ -0,0 +1,27 @@
+            MicronavigationResponse.fromJson(item as Map<String, dynamic>),
+      )
+      .toIList();
+}
```
we could prolly use caching with our `ref.getAndCacheData` helper (see sks menu)

simon-the-shark commented on lib/features/digital_guide_view/tabs/micronavigation/presentation/micronavigation_detail_view.dart:
```diff
@@ -0,0 +1,81 @@
+      const SizedBox(height: DigitalGuideConfig.heightMedium),
+      Text(
+        widget.title,
+        style: TextStyle(
```
we could use `context.textTheme.title.copyWith(fontSize:24)` or smth like this

simon-the-shark commented on lib/features/digital_guide_view/tabs/micronavigation/presentation/micronavigation_detail_view.dart:
```diff
@@ -0,0 +1,81 @@
+
+    return Scaffold(
+      backgroundColor: context.colorTheme.greyLight,
+      appBar: DetailViewAppBar(),
```
don't we wanna accessibility button in this view too in the app bar?

thesun901 replied:
it is pointless on this view, but tbh I forget about it, but I think its good for consistency! Thanks for noticing!

simon-the-shark commented on lib/features/digital_guide_view/tabs/micronavigation/presentation/micronavigation_expansion_tile_content.dart:
```diff
@@ -0,0 +1,84 @@
+          //CHANGE THIS WHEN WE WILL MAKE TRANSLATIONS!
+          onTap: () async {
+            await ref.navigateMicronavigationDetails(
+              response.nameOverride.pl ?? "",
```
tbh, instead of passing every single property by its own, I would consider passing whole `response` as argument

simon-the-shark reviewed (commented):
good job as always :)))

thesun901 commented:
one more thanks for @simon-the-shark also great comments. Sorry guys for being a little bit of Bambik. Some things I've done are really dumb when I look at them now xd... I will try to get the job done before saturday!

simon-the-shark commented:
> one more thanks for @simon-the-shark also great comments. Sorry guys for being a little bit of Bambik. Some things I've done are really dumb when I look at them now xd... I will try to get the job done before saturday!

its actually better code than u probably think it is

thesun901 commented:
Hey guys I applied most of your comments and removed unused fields from model

I also changed name of endpoint in .env, cause this endpoint has more functionalities and it is very probable, that we will use this wrapper again somewhere else

let me know what you think about changes

simon-the-shark reviewed (commented):
veery nicely done. I have few nitpicks, but nothing serious. You can disagree on almost anything I've mentioned if you want

simon-the-shark commented:
you also have one conflict, so a rebase is needed

thesun901 commented on lib/features/digital_guide_view/tabs/micronavigation/data/repository/micronavigation_repository.dart:
```diff
@@ -0,0 +1,39 @@
+          (item) =>
+              MicronavigationResponse.fromJson(item as Map<String, dynamic>),
+        )
+        .toIList(),
```
I will test it and let you know!

thesun901 commented:
I will do the rebase and apply some Simon nitpicks, because I like them!

about this (i dunno why I can answer it directly): "so why can't we have one common MicronavigationTranslations model that is used as type in 2 properties?"

it is separated in api
![image](https://github.com/user-attachments/assets/054d0e07-00df-4f82-9c61-ef4c04df3a3d)

and I don't really know how to connect them using freezed... I will search for that (maybe chat will help me :P), but if you have any idea of references how to do that I will be very thankful

simon-the-shark commented:
> I will do the rebase and apply some Simon nitpicks, because I like them!
> 
> about this (i dunno why I can answer it directly): "so why can't we have one common MicronavigationTranslations model that is used as type in 2 properties?"
> 
> it is separated in api ![image](https://private-user-images.githubusercontent.com/70859223/404460251-054d0e07-00df-4f82-9c61-ef4c04df3a3d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcxNjA0ODgsIm5iZiI6MTczNzE2MDE4OCwicGF0aCI6Ii83MDg1OTIyMy80MDQ0NjAyNTEtMDU0ZDBlMDctMDBkZi00ZjgyLTljNjEtZWY0YzA0ZGYzYTNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE4VDAwMjk0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4NGQ5NjZlZDExYzk4MDQxNmU0ZjE3NDEwNmM1NDJlM2VhMWNiNjEyMmY1YjZlOThiNTMzMWE4MjYzNWNlNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SE_mXuUnP9nCf0bNLXDlh0_NnwQsfYq3JXyP80JIUvU)
> 
> and I don't really know how to connect them using freezed... I will search for that (maybe chat will help me :P), but if you have any idea of references how to do that I will be very thankful

you can just do stmh like this (unless I'm wrong, didn't test it)
```dart
@freezed
class MicronavigationResponse with _$MicronavigationResponse {
  @JsonSerializable(fieldRename: FieldRename.snake)
  const factory MicronavigationResponse({
    required int id,
    required int location,
    required MicronavigationTranslations nameOverride,
    required MicronavigationTranslations webContent,
    required List<MicronavigationLanguage> languages,
  }) = _MicronavigationResponse;

  factory MicronavigationResponse.fromJson(Map<String, dynamic> json) =>
      _$MicronavigationResponseFromJson(json);
}

@freezed
class MicronavigationTranslations with _$MicronavigationTranslations {
  const factory MicronavigationTranslations({
    String? pl,
    String? en,
    String? nb,
    String? de,
    String? uk,
  }) =_MicronavigationTranslations;

  factory MicronavigationTranslations.fromJson(Map<String, dynamic> json) =>
      _$MicronavigationTranslationsFromJson(json);
}
```

thesun901 commented:
I applied every nitpeak of simon! Thanks for the tips on freezed - it works! (and I learned something)... check if you like the final product! And let me know what you think

simon-the-shark reviewed (approved):
didn't test it, but the code lgtm, the rest we'll test in beta

thesun901 commented:
Code didn't give bugs on any availiable buildings or anything, but on beta testing its worth to test how it handles all null values (most null values are behaving as expected but I think there are some that I didnt test, but it is matter of beta testing most functionalities not just micronavigation
